libraries/shared/storage/backfiller.go

```diff
@@ -0,0 +1,165 @@
+		numberOfBins++
+	}
+	blockRangeBins := make([][]uint64, numberOfBins)
+	for i := range blockRangeBins {
+		nextBinStart := bf.startingBlock + uint64(bf.batchSize)
+		if nextBinStart > endingBlock {
+			nextBinStart = endingBlock + 1
+		}
+		blockRange := make([]uint64, 0, nextBinStart-bf.startingBlock+1)
+		for j := bf.startingBlock; j < nextBinStart; j++ {
+			blockRange = append(blockRange, j)
+		}
+		bf.startingBlock = nextBinStart
+		blockRangeBins[i] = blockRange
+	}
+
+	// int64 for atomic incrementing and decrementing to track the number of active processing goroutines we have
+	var activeCount int64
+	// channel for processing goroutines to signal when they are done
+	processingDone := make(chan [2]uint64)
+	forwardDone := make(chan bool)
+
+	// for each block range bin spin up a goroutine to batch fetch and process state diffs for that range
+	go func() {
+		for _, blockHeights := range blockRangeBins {
+			// if we have reached our limit of active goroutines
+			// wait for one to finish before starting the next
+			if atomic.AddInt64(&activeCount, 1) > defaultMaxBatchNumber {
+				// this blocks until a process signals it has finished
+				<-forwardDone
+			}
+			go func(blockHeights []uint64) {
+				payloads, fetchErr := bf.fetcher.FetchStateDiffsAt(blockHeights)
+				if fetchErr != nil {
+					errChan <- fetchErr
+				}
+				for _, payload := range payloads {
+					stateDiff := new(statediff.StateDiff)
+					stateDiffDecodeErr := rlp.DecodeBytes(payload.StateDiffRlp, stateDiff)
+					if stateDiffDecodeErr != nil {
+						errChan <- stateDiffDecodeErr
+						continue
+					}
+					accounts := utils.GetAccountsFromDiff(*stateDiff)
+					for _, account := range accounts {
+						logrus.Trace(fmt.Sprintf("iterating through %d Storage values on account", len(account.Storage)))
+						for _, storage := range account.Storage {
+							diff, formatErr := utils.FromGethStateDiff(account, stateDiff, storage)
+							if formatErr != nil {
+								errChan <- formatErr
+								continue
+							}
+							logrus.Trace("adding storage diff to results",
+								"keccak of address: ", diff.HashedAddress.Hex(),
+								"block height: ", diff.BlockHeight,
+								"storage key: ", diff.StorageKey.Hex(),
+								"storage value: ", diff.StorageValue.Hex())
+							backFill <- diff
+						}
+					}
+				}
+				// when this goroutine is done, send out a signal
+				processingDone <- [2]uint64{blockHeights[0], blockHeights[len(blockHeights)-1]}
+			}(blockHeights)
+		}
+	}()
+
+	// goroutine that listens on the processingDone chan
+	// keeps track of the number of processing goroutines that have finished
+	// when they have all finished, sends the final signal out
+	go func() {
+		goroutinesFinished := 0
+		for {
+			select {
+			case doneWithHeights := <-processingDone:
+				atomic.AddInt64(&activeCount, -1)
+				select {
+				// if we are waiting for a process to finish, signal that one has
+				case forwardDone <- true:
+				default:
+				}
```

Comment:
Could you explain why this is necessary? Why can't we just `forwardDone <- true`?

Reply:
This is necessary because there will not always be a listener on this channel but when there is one we want to send the signal to it. `forwardDone` is only listened to when we are blocked from spinning up the next goroutine [here](https://github.com/vulcanize/vulcanizedb/blob/statediff_at/libraries/shared/storage/backfiller.go#L102) because we have reached the limit. When we reach that limit we wait for a previous goroutine to exit before starting antoher one.